keps/sig-windows/20190424-windows-cri-containerd.md

```diff
@@ -0,0 +1,291 @@
+- runhcs-wcow-process [default] - process isolation is used, container & node OS version must match
+- runhcs-wcow-hypervisor - Hyper-V isolation is used, Pod will be compatible with containers built with Windows Server 2019 / 1809. Physical memory overcommit is allowed with overages filled from pagefile.
+- runhcs-wcow-hypervisor-19H1 - Hyper-V isolation is used, Pod will be compatible with containers built with Windows Server 19H1. Physical memory overcommit is allowed with overages filled from pagefile.
+
+Using Hyper-V isolation does require some extra memory for the isolated kernel & system processes. This could be accounted for by implementing the [PodOverhead](https://kubernetes.io/docs/concepts/containers/runtime-class/#runtime-class) proposal for those runtime classes. We would include a recommended PodOverhead in the default CRDs, likely between 100-200M.
+
+
+#### Improve Control over Memory & CPU Resources with Hyper-V
+
+The Windows kernel itself cannot provide reserved memory for pods, containers or processes. They are always fulfilled using virtual allocations which could be paged out later. However, using a Hyper-V partition improves control over memory and CPU cores. Hyper-V can either allocate memory on-demand (while still enforcing a hard limit), or it can be reserved as a physical allocation up front. Physical allocations may be able to enable large page allocations within that range (to be confirmed) and improve cache coherency. CPU core counts may also be limited so a pod only has certain cores available, rather than shares spread across all cores, and applications can tune thread counts to the actually available cores.
+
+Operators could deploy additional RuntimeClasses with more granular control for performance critical workloads:
+- 2019-Hyper-V-Reserve: Hyper-V isolation is used, Pod will be compatible with containers built with Windows Server 2019 / 1809. Memory reserve == limit, and is guaranteed to not page out.
+  - 2019-Hyper-V-Reserve-<N>Core: Same as above, except all but <N> CPU cores are masked out.
+- 19H1-Hyper-V-Reserve: Hyper-V isolation is used, Pod will be compatible with containers built with Windows Server 19H1. Memory reserve == limit, and is guaranteed to not page out.
+  - 19H1-Hyper-V-Reserve-<N>Core: Same as above, except all but <N> CPU cores are masked out.
+
+
+#### Improved Storage Control with Hyper-V
+
+
+Hyper-V also brings the capability to attach storage to pods using block-based protocols (SCSI) instead of file-based protocols (host file mapping / NFS / SMB). These capabilities could be enabled in HCSv2 with CRI-ContainerD, so this could be an area of future work. Some examples could include:
+
+Attaching a "physical disk" (such as a local SSD, iSCSI target, Azure Disk or Google Persistent Disk) directly to a pod. It could be formatted and used within the pod without being mounted or accessible on the host.
```

Comment:
I understand this is described as future work but I had some questions: In the Hyper-V scenarios, kubelet will continue to run on the host OS and will "see" the disk devices appear first (as the cloud APIs have no visibility into the pods running inside a host). Is the rough idea here to have the HCSv2 enhancements in the kubelet/node plugins "map" the block devices from host up to the appropriate pods running on Hyper-V?

Reply:
ok, clarified